### PR TITLE
GSoC2024: Test Suite - filesystem API

### PR DESCRIPTION
**Pull request by Sadaf-A**
## Changes proposed

- `createDirectory()`
- [x] works for nested directories
- [x] throws error for a duplicate directory name
- [x] throws an error for invalid path names
- `remove()`
- [x] throws an error if the directory does not exist
- [x] removing a file instead of directory works as expected
- `writeFile()`
- [x] works with special characters
- [x] overwrites existing files
- `appendFile()`
- [x] appends to an empty file
- [x] appends an empty string to a file
- [x] appends special characters to a file
- [x] creates and appends to a non-existent file
- [x] appends large content to a fil
- `readFile()`
- [x] reads an empty file
- [x] throws an error if the file does not exist
- `writeBinaryFile()`
- [x] writes a large binary file without throwing errors
- `appendBinaryFile()`
- [x] appends to a non-existent file
- [x] handles concurrent appends
- `readBinaryFile()`
- [x] reads a partially filled file
- [x] reads large files without throwing errors
- `openFile()`
- [x] throws an error when opening a non-existent file
- `updateOpenedFile()`
- [x] throws an error if an invalid action is provided
- [x] handles large data reading in chunks
- [x] throws an error if updating an unopened file
- `getOpenedFileInfo()`
- [x] returns updated pos after reading bytes
- `readDirectory()`
- [x] returns empty array for an empty directory
- [x] returns entries with different file types
- [x] returns entries for nested directories
- [x] throws an error for non-existent directory
- `copy()`
- [x] copies file content correctly
- [x] copies a file into a different directory
- [x] copies a large file correctly
- [x] overwrites the existing file with overwrite option
- [x] skips copying if the destination file exists with skip option
- [x] copies directories recursively with recursive option
- `move()`
- [x] moves a file and retains content
- [x] throws an error when moving to a non-existent directory
- [x] overwrites an existing file
- [x] moves a file to the same location without error
- `getStats()`
- [x] throws an error for non-existent paths
- [x] returns updated modifiedAt after file modification
- [x] returns stats for the file system root
- [x] returns stats for hidden files
- [x] returns stats for empty files

- changed mocha timeout to 22000ms since some tests were failing with 20000ms timeout

## How to test it

 - `cd spec && npm test filesystem`

## Next steps

None.

## Deploy notes

None.

### CODE REVIEW

File: spec/filesystem.spec.js [hunk 1]
```diff
@@ -232,6 +472,34 @@ describe('filesystem.spec: filesystem namespace tests', () => {
...
+                fileIds.push(fileId);
+                fileId = await Neutralino.filesystem.openFile(NL_PATH + '/.tmp/test.txt', 'write');
+                fileIds.push(fileId);
+    
+                await __close(JSON.stringify(fileIds));
+            `);
+            const fileIds = JSON.parse(runner.getOutput());
+            assert.ok(Array.isArray(fileIds));
+            assert.notEqual(fileIds[0], fileIds[1]);
+        });
```
Comment (shalithasuranga) [inline spec/filesystem.spec.js:502]: The `filesystem.openFile()` function only accepts one parameter (filename), so can we can turn this test case into something like "creates file identifiers with incremental integers" (or create any test case by yourself that suites here better) ?

File: spec/filesystem.spec.js [hunk 2]
```diff
@@ -446,6 +923,66 @@ describe('filesystem.spec: filesystem namespace tests', () => {
...
+
+        it('returns stats for the file system root', async () => {
+            runner.run(`
+                let stats = await Neutralino.filesystem.getStats('/');
+                await __close(JSON.stringify(stats));
+            `);
+            let stats = JSON.parse(runner.getOutput());
+            assert.ok(typeof stats == 'object');
+            assert.ok(stats.isDirectory);
+            assert.ok(!stats.isFile);
```
Comment (shalithasuranga) [inline spec/filesystem.spec.js:960]: Could you please do a quick research and verify whether this works properly on Windows (on GitHub Actions also)? Thanks :tada:

File: spec/index.js [hunk 3]
```diff
@@ -11,7 +11,7 @@ fs.readdirSync(testDir).filter((file) => file.includes(specModule + '.spec.js'))
       mocha.addFile(path.join(testDir, file));
 
 });
-mocha.timeout(20000);
+mocha.timeout(22000);
```
Comment (shalithasuranga) [inline spec/index.js:14]: If possible, could you please add some information to this pull request by explaining why you had to increase this timeout? If you can easily find the failing test case for the previous timeout, you can mention it too.

**Review by shalithasuranga — CHANGES_REQUESTED** [review]
Thanks so much for sending this PR. We can merge this after resolving the requested changes. Besides, please check why several test cases are failing (except the `app.exit()` function on Windows -- we need to apply a fix for the framework to fix it I guess). Thanks :tada:

**Comment by Sadaf-A** [comment]
> Thanks so much for sending this PR. We can merge this after resolving the requested changes. Besides, please check why several test cases are failing (except the `app.exit()` function on Windows -- we need to apply a fix for the framework to fix it I guess). Thanks :tada:

Hey, thanks for the review! I'll try to resolve the comments by the EOD

File: spec/filesystem.spec.js [hunk 4]
```diff
@@ -350,18 +666,6 @@ describe('filesystem.spec: filesystem namespace tests', () => {
...
-                await Neutralino.filesystem.writeFile(NL_PATH + '/.tmp/test.txt', 'Hello');
-                let fileId = await Neutralino.filesystem.openFile(NL_PATH + '/.tmp/test.txt');
-                await Neutralino.filesystem.updateOpenedFile(fileId, 'readAll');
-                let info = await Neutralino.filesystem.getOpenedFileInfo(fileId);
-                await __close(JSON.stringify(info));
-            `);
-            let info = JSON.parse(runner.getOutput());
-            assert.equal(info.eof, true);
-        });
-
```
[the remark this answers is not in the document]
Comment (Sadaf-A) [reply spec/filesystem.spec.js]: Hey, I did not intend on removing it I made another commit to add it back but I think it didn't get pushed. I have pushed another commit to add this back. Thanks 🎉

**Review by shalithasuranga — APPROVED** [review]
Thanks so much @Sadaf-A , looks good now :tada:

File: spec/filesystem.spec.js (hunk 2 above)
Comment (shalithasuranga) [reply spec/filesystem.spec.js]: Hi @Sadaf-A, I removed this test case for now. I don't have a Windows machine to experiment with `/`'s behavior on Windows. Please check -- If `/` is a valid path and if it works properly on Windows, you can re-add the following code with another PR:

```js
        it('returns stats for the file system root', async () => {
            runner.run(`
                let stats = await Neutralino.filesystem.getStats('/');
                await __close(JSON.stringify(stats));
            `);
            let stats = JSON.parse(runner.getOutput());
            assert.ok(typeof stats == 'object');
            assert.ok(stats.isDirectory);
            assert.ok(!stats.isFile);
        });
```

I am not sure whether `/` refers to **root** on standard Windows systems.

Thanks :tada:

**Review by shalithasuranga — CHANGES_REQUESTED** [review]
Thanks for sending this @Sadaf-A. I only have one comment regarding these changes. We can merge this after resolving it :tada:

**Pull request merged**